[TELCODOCS-323] - CNF-2772 PTP updates for OpenShift 4.10 (#39713)
openshift/openshift-docs: modules/ztp-configuring-ptp-fast-events.adoc

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: PROCEDURE
[id="ztp-configuring-ptp-fast-events_{context}"]
= Configuring PTP fast events using PolicyGenTemplate custom resources and GitOps ZTP

You can configure PTP fast events for vRAN clusters that are deployed using the GitOps Zero Touch Provisioning (ZTP) pipeline. Use `PolicyGenTemplate` custom resources (CRs) as the basis to create a hierarchy of configuration files tailored to your specific site requirements.

> The `PolicyGenTemplate` CRs that are relevant to PTP events can be found in the `/home/ztp/Argo cd/example` folder in the `quay.io/redhat_emp1/ztp-site-generator:latest` reference architecture container image. The reference architecture has a `/policygentemplates` and `/siteconfig` folder. The `/policygentemplates` folder has common, group, and site-specific configuration CRs. Each `PolicyGenTemplate` CR refers to other CRs that are in the `/source-crs` folder of the reference architecture.

Reviewer: should be /home/ztp/argocd/example

The `PolicyGenTemplate` CRs required to deploy PTP fast events are described below.

.PolicyGenTemplate CRs for vRAN deployments
[cols=2*, options="header"]
|====
|PolicyGenTemplate CR
|Description

|`common-ranGen.yaml`
|Contains the common RAN policies that get applied to all clusters. To deploy the PTP Operator to your clusters, you configure `Namespace`, `Subscription`, and `OperatorGroup` CRs.

|`group-du-3node-ranGen.yaml`
|Contains the RAN policies for three-node clusters only, including PTP fast events configuration.

|`group-du-sno-ranGen.yaml`
|Contains the RAN policies for single-node clusters only, including PTP fast events configuration.

|`group-du-standard-ranGen.yaml`
|Contains the RAN policies for standard three control-plane clusters, including PTP fast events configuration.
|====

.Prerequisites

> * Create a git repository where you manage your custom site configuration data.

Reviewer 2: ```suggestion
* Create a Git repository where you manage your custom site configuration data.
```

Git is capitalized.

* Extract the contents of the `/home/ztp` folder from the `quay.io/redhat_emp1/ztp-site-generator:latest` reference architecture container image, and review the changes.

.Procedure

. Add the following YAML into `.spec.sourceFiles` in the `common-ranGen.yaml` file to configure the AMQP Operator:
+
[source,yaml]
----
#AMQ interconnect operator for fast events
- fileName: AmqSubscriptionNS.yaml
  policyName: "amq-sub-policy"
- fileName: AmqSubscriptionOperGroup.yaml
  policyName: "amq-sub-policy"
- fileName: AmqSubscription.yaml
  policyName: "amq-sub-policy"
----

. Apply the following `PolicyGenTemplate` changes to `group-du-3node-ranGen.yaml`, `group-du-sno-ranGen.yaml`, or `group-du-standard-ranGen.yaml` files according to your requirements:

> .. In `.sourceFiles`, change the `config-policy` to reference the `PtpOperatorConfig` CR that configures the AMQ transport host:

Reviewer: Suggest to use wordings like following, because config-policy is already used by other CRs that are in the ztp example, and given that PtpOperatorConfigForEvent.yaml is not included in default example, we should guide customer to "add" it.
In `.sourceFiles`, add `PtpOperatorConfig` CR that configures the AMQ transport host to `config-policy`:

+
[source,yaml]
----
- fileName: PtpOperatorConfigForEvent.yaml
  policyName: "config-policy"
----

.. Configure the `linuxptp` and `phc2sys` for the PTP clock type and interface. For example, add the following stanza into `.sourceFiles`:
+
[source,yaml]
----
- fileName: PtpConfigSlave.yaml                   <1>

>   policyName: "ptp-config-policy"

Reviewer: We should use "config-policy" instead of "ptp-config-policy" to minimize ztp time. Note that a policy can include multiple CRs as shown in ZTP examples.

  metadata:
    name: "du-ptp-slave"
  spec:
    profile:
    - name: "slave"
      interface: "ens5f1"                         <2>
      ptp4lOpts: "-2 -s --summary_interval -4"    <3>
      phc2sysOpts: "-a -r -m -n 24 -N 8 -R 16"
    ptpClockThreshold:                            <4>
      holdOverTimeout: 30 #secs
      maxOffsetThreshold: 100  #nano secs
      minOffsetThreshold: -100 #nano secs
----
<1> Can be one `PtpConfigMaster.yaml`, `PtpConfigSlave.yaml`, or `PtpConfigSlaveCvl.yaml` depending on your requirements. `PtpConfigSlaveCvl.yaml` configes `linuxptp` services for an Intel E810 Columbiaville NIC.
<2> Device specific interface name.

> <3> You must append the `--summary_interval -4` value to `ptp4lOpts` in `.spec.sourceFiles.spec.profile` to enable PTP fast events. This value must also match the value set for `logSyncInterval` in the corresponding `PtpConfig` CR.

Reviewer: Can we provide example on how to configure logSyncInterval or reference existing examples in other files? e.g., cnf-configuring-cvl-nic-as-oc.adoc has ptp4l configs with logSyncInterval specified.

Author: I've added a to the Columbiaville OC config info right below this content in an `Additional Info` section. LMK what you think. Scroll down to the end of the topic to see: https://deploy-preview-39713--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/ztp-deploying-disconnected.html#ztp-configuring-ptp-fast-events_ztp-deploying-disconnected

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: PROCEDURE
[id="ztp-configuring-ptp-fast-events_{context}"]

> = Configuring PTP fast events using PolicyGenTemplate custom resources and argoCD

Reviewer 3: Maybe:
Configuring PTP fast events using PolicyGenTemplate custom resources and GitOps ZTP

The `PolicyGenTemplate` CRs required to deploy PTP fast events are described below.

.PolicyGenTemplate CRs for vRAN deployments
[cols=2*, options="header"]
|====
|PolicyGenTemplate CR
|Description

|`common-ranGen.yaml`
|Contains the common RAN policies that get applied to all clusters. To deploy the PTP Operator to your clusters, you configure `Namespace`, `Subscription`, and `OperatorGroup` CRs.

|`group-du-3node-ranGen.yaml`
|Contains the RAN policies for three-node clusters only, including PTP fast events configuration.

|`group-du-sno-ranGen.yaml`
|Contains the RAN policies for single-node clusters only, including PTP fast events configuration.

|`group-du-standard-ranGen.yaml`
|Contains the RAN policies for standard three control-plane clusters, including PTP fast events configuration.
|====

.Prerequisites

> * Create a git repo where you manage your custom argoCD site configuration data.

Reviewer 4: ```suggestion
* Create a git repository where you manage your custom argoCD site configuration data.
```

Reviewer 4: s/argoCD/Argo CD (https://argo-cd.readthedocs.io/en/stable/)

> * Create a git repo where you manage your custom argoCD site configuration data.

Reviewer 3: I'd leave argocd out of the description. That is more of an implementation detail. Maybe something like:
Create a git repository where you manage your PolicyGenTemplate configuration data.

* Extract the contents of the `/home/ztp` folder from the `quay.io/redhat_emp1/ztp-site-generator:latest` reference architecture container image, and review the changes.

.Procedure

. Add the following YAML into `.spec.sourceFiles` in the `common-ranGen.yaml` file to configure the AMQP Operator:
+
[source,yaml]
----
#AMQ interconnect operator for fast events
- fileName: AmqSubscriptionNS.yaml
  policyName: "amq-sub-policy"
- fileName: AmqSubscriptionOperGroup.yaml
  policyName: "amq-sub-policy"
- fileName: AmqSubscription.yaml
  policyName: "amq-sub-policy"
----

> . Apply the following `PolicyGenTemplate` changes to `group-du-3node-ranGen.yaml`, `group-du-sno-ranGen.yaml` or `group-du-standard-ranGen.yaml` files according to your requirements:

Reviewer 4: ```suggestion
. Apply the following `PolicyGenTemplate` changes to `group-du-3node-ranGen.yaml`, `group-du-sno-ranGen.yaml`, or `group-du-standard-ranGen.yaml` files according to your requirements:
```

  policyName: "config-policy"
----

.. Configure the `linuxptp` and `phc2sys` for the PTP clock type and interface. For example, add the following stanza into `.sourceFiles`:
+
[source,yaml]
----
- fileName: PtpConfigSlave.yaml                   <1>
  policyName: "ptp-config-policy"
  metadata:
    name: "du-ptp-slave"
  spec:
    profile:
    - name: "slave"
      interface: "ens5f1"                         <2>
      ptp4lOpts: "-2 -s --summary_interval -4"    <3>
      phc2sysOpts: "-a -r -m -n 24 -N 8 -R 16"
    ptpClockThreshold:                            <4>
      holdOverTimeout: 30 #secs
      maxOffsetThreshold: 100  #nano secs
      minOffsetThreshold: -100 #nano secs
----
<1> Can be one `PtpConfigMaster.yaml`, `PtpConfigSlave.yaml`, or `PtpConfigSlaveCvl.yaml` depending on your requirements. `PtpConfigSlaveCvl.yaml` configes `linuxptp` services for an Intel E810 Columbiaville NIC.
<2> Device specific interface name.
<3> You must append the `--summary_interval -4` value to `ptp4lOpts` in `.spec.sourceFiles.spec.profile` to enable PTP fast events. This value must also match the value set for `logSyncInterval` in the corresponding `PtpConfig` CR.

> <4> Optional.

Reviewer 4: Although it's optional, it might help to specify what it does, or how it is used.

> . Merge any other required changes and files with your custom site repo.

Reviewer 4: ```suggestion
. Merge any other required changes and files with your custom site repository.
```

> . Push the changes to your site configuration repo to deploy PTP fast events to new sites using argoCD.

Reviewer 4: ```suggestion
. Push the changes to your site configuration repository to deploy PTP fast events to new sites using Argo CD.
```

> . Push the changes to your site configuration repo to deploy PTP fast events to new sites using argoCD.

Reviewer 3: Maybe use GitOpz ZTP rather than argocd here and below.

. To deploy PTP events to an an existing site, use the Topology-Aware Lifecycle Operator to deploy the changes.

[NOTE]
====

> When the changes are pushed to a site configuration repo, argoCD applies the CRs sequentially according to the `ran.openshift.io/ztp-deploy-wave` annotation. When two or more policies have the same `ran.openshift.io/ztp-deploy-wave` annotation, an alphanumeric sort order is used.

Reviewer 4: ```suggestion
When the changes are pushed to a site configuration repository, Argo CD applies the CRs sequentially according to the `ran.openshift.io/ztp-deploy-wave` annotation. When two or more policies have the same `ran.openshift.io/ztp-deploy-wave` annotation, an alphanumeric sort order is used.
```

> When the changes are pushed to a site configuration repo, argoCD applies the CRs sequentially according to the `ran.openshift.io/ztp-deploy-wave` annotation. When two or more policies have the same `ran.openshift.io/ztp-deploy-wave` annotation, an alphanumeric sort order is used.

Reviewer 3: This note is generally true across any CRs added to the PolicyGenTemplates. Is there a way to reference back to the general GitOps ZTP documentation to avoid duplication?

Author: I moved the note and added a link to the relevant section in the assembly.
